### PR TITLE
Block Editor: Add new `TextAlignmentControl` component

### PR DESCRIPTION
**Pull request by t-hamano**
Part of #60762

## What?

This PR adds a new `TextAlignControl` in preparation for adding block-level text alignment UI to the global styles.

## Why?

Adding UI to the global styles in one PR would involve too many changes, so this PR focuses only on adding new components.

A similar component, the [AlignmentControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/alignment-control), already exists, but since this component is [intended to be used in a toolbar](https://github.com/WordPress/gutenberg/blob/d3f13542a222e906fd6ac2d4992d2516823bac20/packages/block-editor/src/components/alignment-control/ui.js#L52), I think it is necessary to create a new component specifically for it.

## How?

- I referenced the implementation of [TextTransformControl](https://github.com/WordPress/gutenberg/tree/trunk/packages/block-editor/src/components/text-transform-control) which has a similar implementation.
- Add Storybook
- Add README

## Testing Instructions

- Run `npm run storybook:dev` and access `http://localhost:50240/?path=/story/blockeditor-textaligncontrol--default`.
- Confirm that changes to the control property are reflected correctly in the component.
- Confirm that the `onChange` action is called correctly when the button is clicked.
- Check the [README](https://github.com/WordPress/gutenberg/blob/block-editor/add-text-align-control/packages/block-editor/src/components/text-align-control/README.md).

## Screenshots or screencast 

https://github.com/WordPress/gutenberg/assets/54422211/728bb5d8-31d9-4efd-92a3-2b26abe937c0

### CODE REVIEW

**Comment by jasmussen** [comment]
Seems valid insofar as it matches existing patterns for block justification controls. 

On the radar is to use this, or a component like this, to allow a _global_ text justification, including allowing [justified text](https://github.com/WordPress/gutenberg/issues/48202#issuecomment-1976434361). I wonder if this component should be used for that? If yes, I can update the mockup, but also to be aware then that this control should be able to include the justified text toggle.

File: packages/block-editor/README.md
Comment (aaronrobertshaw) [inline packages/block-editor/README.md]:
```suggestion
-   _props.value_ `string`: Currently selected text alignment.
```
Nit: This is an example of the slightly awkward wording with "align" vs "alignment".

The line below reads better to me. Also, making the wording consistent would be a nice tweak.

File: packages/block-editor/src/components/text-align-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-align-control/index.js]: Do you think we need this filtering here given the other text decoration and transform controls didn't?

I believe the display of the others is controlled by theme.json enabling or disabling it. Either way, having them consistent would be good.

On that note, it feels like the three controls could be refactored to just be wrappers for a single control and share some reduced CSS.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for putting this initial step together @t-hamano 👍 

~This is looking pretty good so far. I'd like to wait for a quick sanity check from our design gurus first though to make sure they are happy with the direction.~

_Update: Should of refreshed before submitting the review, we already have it 🚀_ 

The component tests well for me in Storybook and does appear to mostly follow the existing components for text decoration and transform.

https://github.com/WordPress/gutenberg/assets/60436221/a29420d7-c7d4-4869-ac40-408fbeb3f486

I think we might need to tweak the component name, and some of the wording, of docblock comments etc. "Align" is a verb so doesn't make as much sense in some situations.

This is probably why the `AlignmentControl` is named as such. What do you think of renaming this component `TextAlignmentControl`?

**Comment by t-hamano** [comment]
> What do you think of renaming this component `TextAlignmentControl`?

Makes sense, updated by 6c69ce8 👍

File: packages/block-editor/README.md
Comment (t-hamano) [reply packages/block-editor/README.md]: Fixed by b1dfa99 👍

**Comment by t-hamano** [comment]
> On the radar is to use this, or a component like this, to allow a global text justification, including allowing justified text. I wonder if this component should be used for that?

Personally, I find it a bit confusing to indicate text `left` and `justify` with only icons, so I think it's necessary to use a separate component for the global level. I would expect an icon + text UI like the one shown in #48202:

309746977-888c1cab-4e1f-4172-afc7-8fc3df021181

However, since the `TextAlignmentControl` added in this PR is a public component, some users may want to use `justify`. Therefore, I think it makes sense to change this component to the following specifications:

- Default controls are `left`, `center`, and `right`.
- Consumers can also opt-in to `justify` by overriding the `controls` prop.
- If this component is implemented in the global style block level in the future, explicitly exclude `justify`. In other words, even if `[ "left", "center", "justify" ]` is defined in theme.json's `settings.blocks.headings.typography.textAlign`, the buttons that are actually enabled is `left` and `center`.

**Comment by t-hamano** [comment]
> On that note, it feels like the three controls could be refactored to just be wrappers for a single control and share some reduced CSS.

As an alternative approach, I used `Spacer`, and `HStack` components and added a `size="compact"` prop to the `Button` component to avoid writing as much CSS as possible. The layout should be exactly the same as before. Although it is necessary to reset the fieldset element of the component itself, I think we were able to remove most of the CSS.

File: packages/block-editor/src/components/text-align-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-align-control/index.js]: > Do you think we need this filtering here given the other text decoration and transform controls didn't?
> 
> I believe the display of the others is controlled by theme.json enabling or disabling it. Either way, having them consistent would be good.

My understanding is that `TextDecorationControl` and `TextTransformControl` components cannot override controls, so they are guaranteed not to have errors. On the other hand, since this component can overwrite the control, I think an error will occur when the component is used, for example, as shown below.

```jsx
<TextAlignmentControl
  controls={[ 'left', 'center', 'invalid' ]}
  onChange={() => {}}
/>
```

The available values are clearly stated in the documentation, but I personally think it's okay to have this filter.

**Comment by aaronrobertshaw** [comment]
Thanks for iterating on this @t-hamano, I'll give it another review early next week.

> As an alternative approach, I used Spacer, and HStack components and added a size="compact" prop to the Button component to avoid writing as much CSS as possible.

This isn't quite what I meant, or had in mind. It does reduce the CSS for this component but the other two components, text decoration and text transform, have that same CSS as well and the same basic structure.

Take this naming with a grain of salt, but what I was proposing was more like:

- `SegmentedTextControl`: the underlying component for all three text controls; decoration, transform, and alignment
- `TextAlignmentControl`: wrapper for the base component which filters options and passes those items down
- `TextDecorationControl`:  becomes a wrapper, passing options to the base component as well
- `TextTransformControl`: same as the other wrappers

**Comment by t-hamano** [comment]
@aaronrobertshaw Thank you for explaining in detail! I added a new `SegmentedTextControl` component and refactored the component below, which has exactly the same CSS:

- `TextAlignmentControl`
- `TextDecorationControl`
- `TextTransformControl`
- `WritingModeControl` (Not published on Storybook)

Also, while working, I realized that the style sheet for the `WritingModeControl` component (Orientation) was not imported, so it was not displaying as expected. This PR fixes that issue as well.

| trunk | This PR |
|--------|--------|
| image | image |

File: packages/block-editor/src/components/index.js
Comment (t-hamano) [inline packages/block-editor/src/components/index.js]: I didn't add the `__experimental` prefix, does it matter? Or maybe we should expose it as a private component initially.

Comment (noisysocks) [reply packages/block-editor/src/components/index.js]: `__experimental` doesn't mean a lot now that we have to support backwards compatibility for any `__experimental` components that make their way into WP Core.

You could consider making it a private API if you'd like, but I don't think the component you're adding is very controversial or likely to change so I think it's fine to be public from day one.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/segmented-text-control/index.js:38]: Should this use `BaseControl` so that it supports props like `help`? No preference here, just thinking out loud.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/segmented-text-control/index.js]: Would be nice to specify here what attributes each control object should contain.

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (noisysocks) [inline packages/block-editor/src/components/text-alignment-control/README.md]: Is `justify` a valid option?

File: packages/block-editor/src/components/text-alignment-control/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/text-alignment-control/index.js]: Could mark this `string[]`, not `Array`.

File: packages/block-editor/src/components/text-alignment-control/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/text-alignment-control/index.js]: This will create a new array every time, causing `TextAlignmentControl` to always re-render. Is that a problem?

File: packages/block-editor/src/components/writing-mode-control/index.js
Comment (noisysocks) [inline packages/block-editor/src/components/writing-mode-control/index.js]: These need to be changed to `label`.

**Review by noisysocks — COMMENTED** [review]
Looking good! Thanks for refactoring the existing controls, that's a nice simplification.

I think `WritingModeControl` is broken because `controls` needs to pass `label` instead of `name`.

Other than that I've left only minor feedback.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/segmented-text-control/index.js:38]: I was under the impression the `BaseControl` wraps everything in a styled `div` component without passing through additional props so we'd lose the `fieldset`. My memory is failing me but I assume it was required for a11y.

If it was being changed, we'd probably need to remove the `as="legend"` from the `VisualLabel` as well. Maybe such a change would be best left to a follow-up?

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/segmented-text-control/index.js]: Agreed, something mentioning `value`, `icon`, `label` etc. would be handy.

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/text-alignment-control/README.md]: > Consumers can also opt-in to justify by overriding the controls prop.

This [comment](https://github.com/WordPress/gutenberg/pull/60841#issuecomment-2065709938) makes me think it is now and these docs just need updating to the latest.

File: packages/block-editor/src/components/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/index.js]: Updated by 036522b

This component is very simple, so it should be safe to publish it right away, but I've marked it as private just in case, as we might want to tweak it when we incorporate it into the global style.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/segmented-text-control/index.js]: Updated by a050659

File: packages/block-editor/src/components/text-alignment-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-alignment-control/index.js]: Updated by 7439461

File: packages/block-editor/src/components/writing-mode-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/writing-mode-control/index.js]: Fixes by 9f12b6a

File: packages/block-editor/src/components/text-alignment-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-alignment-control/index.js]: Updated by 08fca2a

I applied useMemo hook, does this have any effect?

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/segmented-text-control/index.js:38]: I don't know when such an implementation was first done, but I also think i11y is related. This kind of implementation (fieldset > legend + contents) is done in many components, so I think this is fine for now.

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (t-hamano) [reply packages/block-editor/src/components/text-alignment-control/README.md]: Updated by cc75116 and cfae31a

I've allowed the `justify` value and updated the documentation as well.

image

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/segmented-text-control/index.js]:
```suggestion
 * Control to facilitate selecting a text style from a set of options.
```
To me, this comment doesn't seem to match what this component does.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/segmented-text-control/index.js]: Would `Option` be a more accurate name or description of what these actually are?

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/README.md]:
```suggestion
The `TextAlignmentControl` component is responsible for rendering a control that allows users to select and apply text alignment options to blocks or elements in the Gutenberg editor.
```

Nit: Small tweak to wording to fix the use of the verb `align` as a noun. The second sentence also doesn't add much and sort of overlaps the intent of the first sentence so I think we can remove it.

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/README.md]:
```suggestion
Render the Text Alignment Component with its default `left`, `center`, and `right` options.
```

File: packages/block-editor/src/components/text-alignment-control/README.md
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/README.md]:
```suggestion
The current value of the Text Alignment setting.

-   Type: `String`
-   Required: No
-   Options: `left`, `center`, `right`, `justify`
```

I'm not sure of the preferred approach here but the existing components use `Options` to list the possible values. I think we should be consistent here.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [reply packages/block-editor/src/components/segmented-text-control/index.js]: The more I read through the code in this PR, the more "options" feels like a better fit to me. What do others think?

File: packages/block-editor/src/components/text-alignment-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/index.js]:
```suggestion
 * @param {string[]} props.controls  Array of text alignment controls to display.
```

File: packages/block-editor/src/components/text-alignment-control/stories/index.story.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/stories/index.story.js]: I think we can just import `TextAlignmentControl` directly here. It also then allows a little extra auto-generated documentation in Storybook.

File: packages/block-editor/src/components/text-alignment-control/style.scss
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/style.scss]: Are these styles necessary? Shouldn't the component already have them via the `SegmentedTextControl` styles?

File: packages/block-editor/src/components/text-decoration-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-decoration-control/index.js]: Should we be maintaining the `block-editor-text-decoration-control` class here as well? 

I know it shouldn't be relied upon by consumers but it's only a small tweak to ensure this doesn't change.

File: packages/block-editor/src/components/text-transform-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-transform-control/index.js]: Similar to the text decoration control, I think we should probably maintain the existing `block-editor-text-transform-control` class.

File: packages/block-editor/src/components/writing-mode-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/writing-mode-control/index.js]: We should keep `block-editor-writing-mode-control` just in case consumers did rely on it.

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for your patience and continued effort here @t-hamano 🙇 

I think this PR is moving in the right direction. There's some minor stuff we can clean up but all in all, it's getting close.

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/segmented-text-control/index.js]: Makes sense! Updated by 2288b72

File: packages/block-editor/src/components/text-alignment-control/stories/index.story.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-alignment-control/stories/index.story.js]: Updated by fcb384d

File: packages/block-editor/src/components/text-alignment-control/style.scss
Comment (t-hamano) [reply packages/block-editor/src/components/text-alignment-control/style.scss]: Oh, I forgot to delete this file 😅 Fixed by e038087

File: packages/block-editor/src/components/text-decoration-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-decoration-control/index.js]: Updated by 12f4fc7

File: packages/block-editor/src/components/text-transform-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/text-transform-control/index.js]: Updated by 12f4fc7

File: packages/block-editor/src/components/writing-mode-control/index.js
Comment (t-hamano) [reply packages/block-editor/src/components/writing-mode-control/index.js]: Updated by 12f4fc7

**Review by noisysocks — APPROVED** [review]
Nice work 👍

**Comment by t-hamano** [comment]
@aaronrobertshaw @noisysocks Thank you for your review!

As a next step, I will work on incorporating this UI into the Global Styles for WP6.6. At that time, I would like to consider whether or not to make this component public.

**Comment by t-hamano** [comment]
Just before merging, I noticed that this component has a public README, even though it is private 😅 Do you think the README should also be removed?

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/segmented-text-control/index.js]: Do we need to tweak the `control` vs `option` wording for these as well?

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/segmented-text-control/index.js]:
```suggestion
 * @param {Option[]} props.controls  Array of options to display.
```

File: packages/block-editor/src/components/segmented-text-control/index.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/segmented-text-control/index.js]: It seems like we've sort of gone only halfway to renaming these controls to options.

As I see it, the overall collection of options is the "control", they all set the one style. Each button reflects a different possible value or "option".

**Review by aaronrobertshaw — COMMENTED** [review]
Thanks for the continued refinements here @t-hamano 🙇 

I think we still need to tidy up some of the wording here with regards to"controls" vs "options". With the latest updates we now are a bit inconsistent in the code.

Sorry to be the one holding up this PR landing 😅 

> Do you think the README should also be removed?

Yeah, I believe the current process is to avoid documenting experimental or private components. It's a shame we don't have a mechanism to hide docs or make them private too.

**Comment by t-hamano** [comment]
@aaronrobertshaw Thank you for the additional review 🙇

- I unified JSDoc comments to `options`. I also updated the props name, is this correct?
- Removed README for `TextAlignmentControl`.

File: packages/block-editor/src/components/text-alignment-control/stories/index.story.js
Comment (aaronrobertshaw) [inline packages/block-editor/src/components/text-alignment-control/stories/index.story.js]:
```suggestion
		options: {
```
Missed tweaking the story prop name when switching. from controls to options 🙂

**Review by aaronrobertshaw — APPROVED** [review]
Thanks for the persistence here @t-hamano 👍 

Sorry, this PR got a little more complicated than it started out but I think we've arrived at a good place with it.

Each of the controls are working well for me in the editor. There's just one small thing that was missed with the storybook example for the new control. Once that's updated this should be good to merge.

**Comment by t-hamano** [comment]
@aaronrobertshaw Thanks for the review!

I fixed the storybook control name. Since I have confirmed that it works correctly on Storybook, I would like to merge it 👍

**Comment by aaronrobertshaw** [comment]
> I would like to merge it 

Go for it! Thanks again for the hard work here 👍

**Pull request merged**
